odl/space/space_utils.py

```diff
@@ -203,6 +206,233 @@ def rn(size, dtype=None, impl='numpy', **kwargs):
     return rn
 
 
+class auto_weighting(OptionalArgDecorator):
+
+    """Make an unweighted adjoint automatically account for weightings.
+
+    Depending on the weightings, the correction is achieved by composing
+    the unweighted operator with either `ScalingOperator` or
+    `ConstantOperator`. The following rules are applied for the domain
+    weighting ``w``, the range weighting ``v`` and the provided unweighted
+    adjoint ``B^*``:
+
+    - If both ``w`` and ``v`` are arrays, return ::
+
+        (1 / w) * (B^*) * v
+
+    - If ``w`` is an array and ``v`` a constant, return ::
+
+        (v / w) * (B^*)
+
+    - If ``w`` is a constant and ``v`` an array, return ::
+
+        (B^*) * (w / v)
+
+    - If both ``w`` and ``v`` are constants, return ::
+
+        (B^*) * (v / w)
+
+      if ``B.range.size < B.domain.size``, otherwise ::
+
+        (v / w) * (B^*)
+
+    - Ignore constants 1.0.
+
+    To avoid the inconvenience of dealing with `OperatorComp` objects,
+    the given operator is monkey-patched instead of composed.
+
+    Parameters
+    ----------
+    unweighted_adjoint : `Operator`
+        Unweighted variant of the adjoint. It will be patched with a
+        new ``_call()`` method.
+        The weightings of ``domain`` and ``range`` of the operator
+        must be `ArrayWeighting` or `ConstWeighting`.
```

Comment:
and we still need to somehow deal with `NoWeighting`

Reply:
I don't see that. Unweighted is whatever comes in as adjoint operator, and all the other weightings are defined relative to that.